meta.go

```diff
@@ -0,0 +1,100 @@
+package yarpc
+
+import (
+	"context"
+
+	"go.uber.org/yarpc/api/transport"
+)
+
+// CallOption defines options that may be passed in at call sites to other
+// services.
+//
+// These may be used to add or alter the request.
+type CallOption struct{ apply func(*Call) }
+
+type callHeader struct{ k, v string }
+
+// Call represents an outgoing call.
+//
+// It holds any per-call options for a request. Encoding authors may use Call
+// to hydrate Requests from call-site options.
+type Call struct {
+	// request attributes to fill if non-nil
+	headers         []callHeader
+	shardKey        *string
+	routingKey      *string
+	routingDelegate *string
+
+	// If non-nil, response headers should be written here.
+	responseHeaders *Headers
+}
+
+// WriteToRequest fills the given request with request-specific options from
+// the call.
+func (c *Call) WriteToRequest(ctx context.Context, req *transport.Request) {
+	for _, h := range c.headers {
+		req.Headers = req.Headers.With(h.k, h.v)
+	}
+
+	if c.shardKey != nil {
+		req.ShardKey = *c.shardKey
+	}
+	if c.routingKey != nil {
+		req.RoutingKey = *c.routingKey
+	}
+	if c.routingDelegate != nil {
+		req.RoutingDelegate = *c.routingDelegate
+	}
+}
+
+// ReadFromResponse reads information from the response for this call.
+//
+// This should be called only if the request is unary.
+func (c *Call) ReadFromResponse(ctx context.Context, res *transport.Response) {
+	// We're not using ctx right now but we may in the future.
+	if c.responseHeaders != nil {
+		*c.responseHeaders = Headers(res.Headers)
+	}
+}
+
+// ApplyCallOptions applies options to the given Call.
+func ApplyCallOptions(call *Call, options ...CallOption) {
+	for _, opt := range options {
+		opt.apply(call)
+	}
+}
+
+// ResponseHeaders specifies that headers received in response to this request
+// should be fed into the given object.
+//
+// 	var resHeaders yarpc.Headers
+// 	resBody, err := client.SetValue(ctx, key, value, yarpc.ResponseHeaders(&resHeaders))
+func ResponseHeaders(h *Headers) CallOption {
+	return CallOption{func(o *Call) { o.responseHeaders = h }}
+}
+
+// TODO(abg): Example tests
+
+// WithHeader adds a new header to the request.
+//
+// 	resBody, err := client.GetValue(ctx, reqBody, yarpc.WithHeader("Token", "10"))
+func WithHeader(k, v string) CallOption {
```

Comment:
Looks like you want headers as an option, not attached to the context.

I'm ok to go that path, and if we end up not liking it, we can course correct within 1.x by deprecating this and introducing the "ride the context" API.